### DISCUSSION TITLE
On logging

### DISCUSSION THREAD

**Discussion by pimeys** [napi-rs/napi-rs#471] [Q&A]
We're trying to think about logging and how we can operate between Rust and JavaScript. What we're after here is our Rust code produces structured logs with `tracing`, and we'd like our JavaScript side to be able to catch whatever gets written to STDOUT with a callback, make something with the data and decide on writing to the output. I'd like to understand a bit about the best ways of doing this with napi.rs. Beforehand we had a separate process and we could have a pointer to the process output, but now we're int the same process so we might need to do something smarter.

**Comment by Brooooooklyn** [comment]
I need to get familiar with the usage of tracing first, and this week I will try to create a sample project to achieve this requirement as you described.

Happy Chinese new year 🎉

**Reply by pimeys** [reply]
Happy new year!

**Comment by pimeys** [comment]
Here's an example how to stream logs from Rust to JavaScript. Not the most efficient way of doing it (you should really log in Rust). But, if you need to do this, it could work a bit like this:

https://github.com/pimeys/napi-events

**Reply by Brooooooklyn** [reply]
I'm working on https://github.com/pimeys/napi-events/pull/1, logging with `ThreadSafeFunction` seems to be more reasonable, but I still exploring how to shutdown the subscriber properly.

**Comment by pimeys** [comment]
One more question about this. When calling the `ThreadsafeFunction`, is it safe to use the `NonBlocking` call mode? N-API docs recommend not using it from the same thread with the JavaScript interpreter, but in our case we're always spawned to a tokio pool, e.g. in a different thread. I'm a bit worried a blocking call here would slow down the tokio reactor.

**Reply by Brooooooklyn** [reply]
I believe it’s safe to call TSFN with NonBlocking call mode in this scenario
